sc2/game_data.py

```diff
@@ -2,322 +2,378 @@
 from sc2.ids.unit_typeid import UnitTypeId
-from sc2.unit_command import UnitCommand
 
 # Set of parts of names of abilities that have no cost
 # E.g every ability that has 'Hold' in its name is free
 # TODO move to constants, add more?
 FREE_ABILITIES = {"Lower", "Raise", "Land", "Lift", "Hold", "Harvest"}
 
 
+@dataclass
 class GameData:
+    _abilities: List[Any] = field(default_factory=list)
+    _units: List[Any] = field(default_factory=list)
+    _upgrades: List[Any] = field(default_factory=list)
+    _buffs: List[Any] = field(default_factory=list)
+    _effects: List[Any] = field(default_factory=list)
+
+    _ability_cost_cache: ClassVar[Dict[int, Cost]] = {}
 
-    def __init__(self, data):
-        """
-        :param data:
-        """
-        ids = set(a.value for a in AbilityId if a.value != 0)
-        self.abilities: Dict[int, AbilityData] = {
-            a.ability_id: AbilityData(self, a)
-            for a in data.abilities if a.ability_id in ids
+    @classmethod
+    def from_proto(cls, data: Any):
+        data_as_dict = MessageToDict(data, including_default_value_fields=False, preserving_proto_field_name=True)
+        underscore_prepended = {f"_{key}": value for key, value in data_as_dict.items()}
+        return GameData(**underscore_prepended)
+
+    @cached_property
+    def ids(self) -> Set[Any]:
+        return set(a.value for a in AbilityId if a.value != 0)
+
+    @cached_property
+    def abilities(self) -> Dict[int, AbilityData]:
+        return {
+            a['ability_id']: AbilityData(**a, _game_data=self)
+            for a in self._abilities if a['ability_id'] in self.ids
         }
-        self.units: Dict[int, UnitTypeData] = {u.unit_id: UnitTypeData(self, u) for u in data.units if u.available}
-        self.upgrades: Dict[int, UpgradeData] = {u.upgrade_id: UpgradeData(self, u) for u in data.upgrades}
-        # Cached UnitTypeIds so that conversion does not take long. This needs to be moved elsewhere if a new GameData object is created multiple times per game
 
-    @lru_cache(maxsize=256)
-    def calculate_ability_cost(self, ability: Union[AbilityData, AbilityId, UnitCommand]) -> Cost:
-        if isinstance(ability, AbilityId):
-            ability = self.abilities[ability.value]
-        elif isinstance(ability, UnitCommand):
-            ability = self.abilities[ability.ability.value]
+    @cached_property
+    def units(self) -> Dict[int, UnitTypeData]:
+        return {u['unit_id']: UnitTypeData(**u, _game_data=self) for u in self._units if u['available']}
 
-        assert isinstance(ability, AbilityData), f"Ability is not of type 'AbilityData', but was {type(ability)}"
+    @cached_property
+    def upgrades(self) -> Dict[int, UpgradeData]:
+        return {u['upgrade_id']: UpgradeData(**u, _game_data=self) for u in self._upgrades}
 
-        for unit in self.units.values():
-            if unit.creation_ability is None:
-                continue
+    # TODO: Buffs and effects from GameData
```

Comment:
Does this still need to be done?

Reply:
They can be added (in this PR), but they weren't used before, and I dont know if they will be useful at all
![image](https://user-images.githubusercontent.com/21288725/194569760-2c3be37a-1b5d-4d70-9de5-8ca460bcbbe9.png)
![image](https://user-images.githubusercontent.com/21288725/194569647-63c4b9ae-842e-4721-a44e-a126c8f1b3f7.png)